sqs/user_template_test.go

```diff
@@ -106,101 +106,114 @@ var _ = Describe("UserTemplate", func() {
-				"sqs:ChangeMessageVisibility",
-				"sqs:DeleteMessage",
-				"sqs:GetQueueAttributes",
-				"sqs:GetQueueUrl",
-				"sqs:ListDeadLetterSourceQueues",
-				"sqs:ListQueueTags",
-				"sqs:PurgeQueue",
-				"sqs:ReceiveMessage",
-				"sqs:SendMessage",
-			))
+			Expect(policy.PolicyDocument).To(
+				HaveKeyWithValue("Statement", ConsistOf(
+					HaveKeyWithValue("Action", ConsistOf(
+						"sqs:ChangeMessageVisibility",
+						"sqs:DeleteMessage",
+						"sqs:GetQueueAttributes",
+						"sqs:GetQueueUrl",
+						"sqs:ListDeadLetterSourceQueues",
+						"sqs:ListQueueTags",
+						"sqs:PurgeQueue",
+						"sqs:ReceiveMessage",
+						"sqs:SendMessage",
+					)),
+				)))
 		})
 	})
 
 	It("should return an error for unknown access policies", func() {
-		t, err := sqs.UserTemplate(sqs.UserParams{
+		t, err := sqs.UserTemplateBuilder{
 			AccessPolicy: "bananas",
-		})
-		Expect(t).To(BeNil())
+		}.Build()
+
+		Expect(t).To(BeEmpty())
 		Expect(err).To(MatchError("unknown access policy \"bananas\""))
 	})
 
 	Context("when queue ARNs are set", func() {
 		BeforeEach(func() {
-			params.PrimaryQueueARN = "abc"
-			params.SecondaryQueueARN = "qwe"
+			builder.PrimaryQueueARN = "abc"
+			builder.SecondaryQueueARN = "qwe"
 		})
 		It("the policy is scoped to the queue ARNs", func() {
-			Expect(policy.PolicyDocument).To(BeAssignableToTypeOf(sqs.PolicyDocument{}))
-			policyDoc := policy.PolicyDocument.(sqs.PolicyDocument)
-			Expect(policyDoc.Statement).To(HaveLen(1))
-			Expect(policyDoc.Statement[0].Effect).To(Equal("Allow"))
-			Expect(policyDoc.Statement[0].Action).To(ConsistOf(
-				"sqs:ChangeMessageVisibility",
-				"sqs:DeleteMessage",
-				"sqs:GetQueueAttributes",
-				"sqs:GetQueueUrl",
-				"sqs:ListDeadLetterSourceQueues",
-				"sqs:ListQueueTags",
-				"sqs:PurgeQueue",
-				"sqs:ReceiveMessage",
-				"sqs:SendMessage",
-			))
-			Expect(policyDoc.Statement[0].Resource).To(ConsistOf("abc", "qwe"))
+			Expect(policy.PolicyDocument).To(
+				HaveKeyWithValue("Statement", ConsistOf(
+					And(
+						HaveKeyWithValue("Effect", "Allow"),
+						HaveKeyWithValue("Resource", ConsistOf("abc", "qwe")),
+						HaveKeyWithValue("Action", ConsistOf(
+							"sqs:ChangeMessageVisibility",
+							"sqs:DeleteMessage",
+							"sqs:GetQueueAttributes",
+							"sqs:GetQueueUrl",
+							"sqs:ListDeadLetterSourceQueues",
+							"sqs:ListQueueTags",
+							"sqs:PurgeQueue",
+							"sqs:ReceiveMessage",
+							"sqs:SendMessage",
+						))),
+				)))
 		})
 	})
 
 	It("should create an active access key", func() {
-		Expect(accessKey.Status).To(Equal("Active"))
-		Expect(accessKey.UserName).ToNot(BeEmpty())
+		var result map[string]interface{}
+		Expect(yaml.Unmarshal([]byte(rawText), &result)).To(Succeed())
+		resources := result["Resources"].(map[interface{}]interface{})
+		resource := resources[sqs.ResourceAccessKey].(map[interface{}]interface{})
+		properties := resource["Properties"].(map[interface{}]interface{})
```

Comment:
Don't these casts give you an `ok` return value too?

Reply:
this form (without `ok`) will panic if the assertion does not hold, which is fine in this context.  Usually we use the `ok` form if we want to do custom error handling rather than just panicking.

https://tour.golang.org/methods/15

Reply:
Gotcha.